release-20.2: rowexec: account for some allocations due to hashing of datums

yuzefovich opened the pull request:
Backport 1/1 commits from #54770.

/cc @cockroachdb/release

---

This commit adds memory accounting for the allocations of `tree.Datum`s
that might occur during `Fingerprint`ing (for example, when the encoded
datum uses a value encoding and we want to use the key ASC encoding). In
many cases, the newly allocated datums are only needed temporary, but we
know that GC takes its time with reclaiming the memory, so we choose to
go this route of possibly over-accounting in several places (in case of
the hash aggregator, windower, and sampler). This should improve the
stability of CockroachDB by preventing some crashes.

Note that the sampler is using an unlimited memory account for the
tracking of these allocations because if we limit it with default 64MB,
the stats jobs might never succeed and the throttling should kick in to
prevent the stats collection to reserve all the memory in the cluster.

Addresses: #54360.
Informs: #54670.

Release note: None

yuzefovich committed:
rowexec: account for some allocations due to hashing of datums

This commit adds memory accounting for the allocations of `tree.Datum`s
that might occur during `Fingerprint`ing (for example, when the encoded
datum uses a value encoding and we want to use the key ASC encoding). In
many cases, the newly allocated datums are only needed temporary, but we
know that GC takes its time with reclaiming the memory, so we choose to
go this route of possibly over-accounting in several places (in case of
the hash aggregator, windower, distinct, and sampler). This should
improve the stability of CockroachDB by preventing some crashes.

Note that the sampler is using an unlimited memory account for the
tracking of these allocations because if we limit it with default 64MB,
the stats jobs might never succeed and the throttling should kick in to
prevent the stats collection to reserve all the memory in the cluster.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/55145)
<!-- Reviewable:end -->

yuzefovich commented:
Will wait for 20.2.1 to merge.

yuzefovich committed:
rowexec: remove recently added memory accounting in sampler

We recently added the memory accounting for `tree.Datum`s that could be
allocated during stats collection during fingerprinting. The idea was
that Go GC is slow, so it's a good idea to account for those allocations
even though the references to the datums are lost right away. After
thinking about it more and doing some experiments, I think it's a bad
idea because it could lead to stats jobs hitting a memory budget error
in cases when it shouldn't happen which would result in auto stats never
completing. We will just keep our old behavior of possibly running into
OOMs until we implement the vectorized stats collection and can make
easier improvements with memory accounting.

Release note: None

yuzefovich commented:
I also changed my mind on including stats change.

yuzefovich commented:
Closing in favor of #55380.
